src/porepy/numerics/ad/_ad_utils.py

```diff
@@ -447,6 +447,51 @@ def get_solution_values(
     return value
 
 
+def shift_solution_values(
+    name: str,
+    data: dict,
+    location: Any,
+) -> None:
+    """Function to shift numerical values stored in the data dictionary.
+
+    The shift is implemented s.t. values at index ``i`` are copied to index ``i + 1``,
+    based on the number of stored values.
+
+    Note:
+        The data stored must have support for ``.copy()`` in order to avoid faulty
+        referencing (e.g., numpy arrays or sparse matrices).
+
+    Parameters:
+        name: Key in ``data`` for which quantity the shift should be performed.
+        data: A grid data dictionary.
+        location: Either :data:`~porepy.utils.common_constants.TIME_STEP_SOLUTIONS`
+            or :data:`~porepy.utils.common_constants.ITERATE_SOLUTIONS`.
+
+    Raises:
+        ValueError: If unsupported ``location`` is passed.
+
+    """
+    if location not in data:
+        data[location] = {}
+    if name not in data[location]:
+        # TODO should we return here instead? Nothing to be shifted, and
+        # set_solution_values is the method which prepares the data dict
+        data[location][name] = {}
+
+    num_stored = len(data[location][name])
+
+    if location == pp.ITERATE_SOLUTIONS:
+        range_ = range(num_stored, 0, -1)
```

Comment:
@keileg in the original function in EquationSystem, the range for the case `pp.ITERATE_SOLUTIONS` was `range(num_stored - 1, 0, -1)`.
I don't know if I am introducing a breaking change or if this is a fix (no test coverage), but the original implementation had the (I think unwanted) consequence, that iterate values where not shifted, if no previous iterate was set (index > 0) (so values at index 0 were never shifted to index 1)